zrml/parimutuel/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+                market_id,
+                buyer: who,
+                asset,
+                amount_minus_fees,
+                fees: external_fees,
+            });
+
+            Ok(())
+        }
+
+        fn do_claim_rewards(who: T::AccountId, market_id: MarketIdOf<T>) -> DispatchResult {
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Resolved, Error::<T>::MarketIsNotResolvedYet);
+            ensure!(market.scoring_rule == ScoringRule::Parimutuel, Error::<T>::InvalidScoringRule);
+            ensure!(
+                matches!(market.market_type, MarketType::Categorical(_)),
+                Error::<T>::NotCategorical
+            );
+            let winning_outcome = market.resolved_outcome.ok_or(Error::<T>::NoResolvedOutcome)?;
+            let pot_account = Self::pot_account(market_id);
+            let winning_asset = match winning_outcome {
+                OutcomeReport::Categorical(category_index) => {
+                    Asset::ParimutuelShare(market_id, category_index)
+                }
+                OutcomeReport::Scalar(_) => return Err(Error::<T>::NotCategorical.into()),
+            };
+            // each Parimutuel outcome asset has the market id included
+            // this allows us to query all outstanding shares for each discrete asset
+            let outcome_total = T::AssetManager::total_issuance(winning_asset);
+            // if there are no outstanding reward shares, but the pot account is not empty
+            // then use the refund extrinsic to get the initial bet back
+            ensure!(outcome_total != BalanceOf::<T>::zero(), Error::<T>::NoRewardShareOutstanding);
+            let winning_balance = T::AssetManager::free_balance(winning_asset, &who);
+            ensure!(!winning_balance.is_zero(), Error::<T>::NoWinningShares);
+            debug_assert!(
+                outcome_total >= winning_balance,
+                "The outcome issuance should be at least as high as the individual balance of \
+                 this outcome!"
+            );
+
+            let pot_total = T::AssetManager::free_balance(market.base_asset, &pot_account);
+            let payoff_ratio_mul_base: BalanceOf<T> =
+                bdiv_floor(pot_total.saturated_into(), outcome_total.saturated_into())?
+                    .saturated_into();
+            let payoff: BalanceOf<T> = bmul_floor(
+                payoff_ratio_mul_base.saturated_into(),
+                winning_balance.saturated_into(),
+            )?
+            .saturated_into();
+
+            Self::check_values(
+                winning_balance,
+                pot_total,
+                outcome_total,
+                payoff_ratio_mul_base,
+                payoff,
+            )?;
+
+            let slashable_asset_balance = winning_balance;
+
+            T::AssetManager::slash(winning_asset, &who, slashable_asset_balance);
```

Comment:
Slash is generally an operation that is [associated with punishment](https://substrate.stackexchange.com/questions/2379/about-some-balance-terminology/2382#2382). Furthermore, without guards it could also [reduce the reserved balance](https://github.com/open-web3-stack/open-runtime-module-library/blob/b3694e631df7f1ca16b1973122937753fcdee9d4/tokens/src/lib.rs#L1189-L1190) in addition to the free balance (doesn't matter here, but you never know how code develops).

If you simply want to reduce the account balance, especially if it is not related to punishment, using `withdraw` is recommended.

```suggestion
            T::AssetManager::withdraw(winning_asset, &who, winning_balance)?;
```

Reply:
098f17faf4abeed16866d7fa7e7f52177b0bbbef